### ISSUE TITLE
Deno lacks a cookie jar for `fetch`

### ISSUE THREAD

**Issue by legowerewolf** [denoland/deno#5862]
Deno should store cookies on a per-run basis (and optionally have methods in the Deno namespace for export/import of cookies) so that it's possible to make a series of requests that depend on cookies. 

Right now, a flow like the following would be impossible:

1. No cookies sent; response sets cookie 1.
2. Cookie 1 sent; response uses cookie 1 value somewhere in the body

Yes, I'm aware of `std/http/cookie`. No, it's not what I need - this is not a server. It's a client.

**Comment by ry** [comment]
I agree. Deno ought to behave exactly as a browser does when it comes to the `fetch()` API.

**Comment by marcosc90** [comment]
> Right now, a flow like the following would be impossible:

it's currently possible to implement a cookie jar on userland.

```javascript
const res = await fetch('https://example.com/set-cookie');
const cookies = res.headers.get('set-cookie')
// Implement your jar logic
const headers = new Headers();
headers.set('Cookie', cookies); // get from jar
const res = await fetch('https://example.com/set-cookie', { headers });

**Comment by legowerewolf** [comment]
If Deno's `fetch` and `headers` APIs match specs, the Cookie header is one of the [forbidden header names](https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name), and that won't work. I'm surprised that that library works after inspecting it.

**Comment by marcosc90** [comment]
It's forbidden on the client, it wouldn't make sense to make that limitation on a server side HTTP client to be honest. 

Deno should match the spec where it makes sense to match it. The same way we had to go around the spec for `Set-Cookie` header.

**Comment by legowerewolf** [comment]
I fiddled around a bit with Deno and Postman-Echo. Deno sends cookie headers. I must be missing something. (I'm trying to rewrite https://github.com/legowerewolf/ao3-fetch-js for Deno and running into snags in the login process.)

**Comment by isuckatprogram** [comment]
I have the same issue. I am using Deno to send cookie headers and when I test it on postman I can see the cookie header but for some reason, the cookie doesn't work.

**Comment by ashersaupingomez** [comment]
I have the same issue as well. Lack of cookie support for the fetch API is the main reason I haven't been able to move over to deno. This functionality would be great to have!

**Comment by jd1378** [comment]
Hi
I made one, you can try it out:
https://deno.land/x/another_cookiejar@v1.0.0

**Comment by shivam-tripathi** [comment]
Parsing cookie in result of fetch is a big pain - res.headers.get('cookie') return `;` separated not just key value but also things like:
```
Path=/; Secure; SameSite=None, <key>=<value>;Expires=Sat, 26 Nov 2022 17:31:13 GMT; 
```
Because of SameSite it's even difficult to extract key, value from `fetch(...).then(res => res.headers.get('cookie'))`. Not sure how to parse cookie to extract out all key value.

**Comment by bartlomieju** [comment]
@shivam-tripathi use `cookie` module from `deno_std` https://deno.land/std@0.116.0/http/cookie.ts

**Comment by shivam-tripathi** [comment]
@bartlomieju I am guessing you want to point to [getCookies](https://deno.land/std@0.116.0/http/cookie.ts#L170) function in  [https://deno.land/std@0.116.0/http/cookie.ts](https://deno.land/std@0.116.0/http/cookie.ts).
Unfortunately, this method will not work - because it assume there are only key value pairs separated by semicolons. In fetch's response cookies contain other fields like Secure, Expires etc all of them separated by semi colon as well - which breaks this method. Moreover, all the cookies are comma separated, not semicolon separated.

**Comment by bartlomieju** [comment]
@shivam-tripathi yes I meant `getCookies`. @lucacasonato what do you think?

**Comment by lucacasonato** [comment]
@shivam-tripathi Are you talking about `set-cookie`, or `cookie`. Fields like `Expires`, `Secure` are present in `set-cookie`, but not `cookie`. To get an unconcatenated list of all `set-cookie` headers, do the following:

```js
const setCookieHeaders = [...headers.entries()].filter(([k]) => k === 'set-cookie').map(([k, v]) => v);
```

**Comment by shivam-tripathi** [comment]
@lucacasonato Hey, yes you're right I meant `set-cookie` not `cookie` - res.headers.get('cookie') gives `null`. Using `headers.entries()` to filter out all `set-cookie` headers solved it for me. 
Thanks @bartlomieju @lucacasonato!

**Comment by garretmh** [comment]
@shivam-tripathi The behavior of Headers#get() is defined by WHATWG.

> headers . get(name)
> Returns as a string the values of all headers whose name is name, separated by a comma and a space.

<https://fetch.spec.whatwg.org/#ref-for-dom-headers-get①>

**Comment by iuioiua** [comment]
One can now use the following as a workaround:
```ts
import makeFetchCookie from "npm:fetch-cookie";

const fetchCookie = makeFetchCookie(fetch);
```

**Comment by iuioiua** [comment]
What are the current standings, from both the Deno team and the community, on supporting automatic cookie handling in the Fetch API?

Personally, I'd like to have it implemented, same as the browser and within Deno to ensure correctness and quality. However, it'd come with added complexity and security concerns.

AFAIK, no other server-side JS runtime has this functionality native. There could be good reasons for this, as previously mentioned, but this could be an opportunity for Deno 👀

**Comment by phil294** [comment]
> I agree. Deno ought to behave exactly as a browser does when it comes to the fetch() API.

so... shouldn't Deno introduce a new `document` namespace then so we can do `document.cookie = 'a=b;'; fetch(...)`? As that *is* how the browser behaves

**Comment by xjfnet** [comment]
https://deno.land/x/another_cookiejar/mod.ts

This library offers a fetch wrapper that retain cookies. This library also provides a simple cookiejar;

**Comment by garretmh** [comment]
Could an option be added to `Deno.createHttpClient()` to enable cookie handling?
